File: swift/Sources/MaterialColorUtilities/Scheme/Scheme.swift
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

@available(*, deprecated, message: "Use `DynamicScheme`")
class Scheme {
  let primary: Int
  let onPrimary: Int
  let primaryContainer: Int
  let onPrimaryContainer: Int
  let secondary: Int
  let onSecondary: Int
  let secondaryContainer: Int
  let onSecondaryContainer: Int
  let tertiary: Int
  let onTertiary: Int
  let tertiaryContainer: Int
  let onTertiaryContainer: Int
  let error: Int
  let onError: Int
  let errorContainer: Int
  let onErrorContainer: Int
  let background: Int
  let onBackground: Int
  let surface: Int
  let onSurface: Int
  let surfaceVariant: Int
  let onSurfaceVariant: Int
  let outline: Int
  let outlineVariant: Int
  let shadow: Int
  let scrim: Int
  let inverseSurface: Int
  let inverseOnSurface: Int
  let inversePrimary: Int

  init(
    primary: Int, onPrimary: Int, primaryContainer: Int, onPrimaryContainer: Int, secondary: Int,
    onSecondary: Int, secondaryContainer: Int, onSecondaryContainer: Int, tertiary: Int,
    onTertiary: Int, tertiaryContainer: Int, onTertiaryContainer: Int, error: Int, onError: Int,
    errorContainer: Int, onErrorContainer: Int, background: Int, onBackground: Int, surface: Int,
    onSurface: Int, surfaceVariant: Int, onSurfaceVariant: Int, outline: Int, outlineVariant: Int,
    shadow: Int, scrim: Int, inverseSurface: Int, inverseOnSurface: Int, inversePrimary: Int
  ) {
    self.primary = primary
    self.onPrimary = onPrimary
    self.primaryContainer = primaryContainer
    self.onPrimaryContainer = onPrimaryContainer
    self.secondary = secondary
    self.onSecondary = onSecondary
    self.secondaryContainer = secondaryContainer
    self.onSecondaryContainer = onSecondaryContainer
    self.tertiary = tertiary
    self.onTertiary = onTertiary
    self.tertiaryContainer = tertiaryContainer
    self.onTertiaryContainer = onTertiaryContainer
    self.error = error
    self.onError = onError
    self.errorContainer = errorContainer
    self.onErrorContainer = onErrorContainer
    self.background = background
    self.onBackground = onBackground
    self.surface = surface
    self.onSurface = onSurface
    self.surfaceVariant = surfaceVariant
    self.onSurfaceVariant = onSurfaceVariant
    self.outline = outline
    self.outlineVariant = outlineVariant
    self.shadow = shadow
    self.scrim = scrim
    self.inverseSurface = inverseSurface
    self.inverseOnSurface = inverseOnSurface
    self.inversePrimary = inversePrimary
  }

  static func light(_ color: Int) -> Scheme {
    return lightFromCorePalette(CorePalette.of(color))
  }

  static func lightContent(_ color: Int) -> Scheme {
    return lightFromCorePalette(CorePalette.contentOf(color))
  }

  static func dark(_ color: Int) -> Scheme {
    return darkFromCorePalette(CorePalette.of(color))
  }

  static func darkContent(_ color: Int) -> Scheme {
    return darkFromCorePalette(CorePalette.contentOf(color))
  }

  static func lightFromCorePalette(_ palette: CorePalette) -> Scheme {
    return Scheme(
      primary: palette.primary.tone(40),
      onPrimary: palette.primary.tone(100),
      primaryContainer: palette.primary.tone(90),
      onPrimaryContainer: palette.primary.tone(10),
      secondary: palette.secondary.tone(40),
      onSecondary: palette.secondary.tone(100),
      secondaryContainer: palette.secondary.tone(90),
      onSecondaryContainer: palette.secondary.tone(10),
      tertiary: palette.tertiary.tone(40),
      onTertiary: palette.tertiary.tone(100),
      tertiaryContainer: palette.tertiary.tone(90),
      onTertiaryContainer: palette.tertiary.tone(10),
      error: palette.error.tone(40),
      onError: palette.error.tone(100),
      errorContainer: palette.error.tone(90),
      onErrorContainer: palette.error.tone(10),
      background: palette.neutral.tone(99),
      onBackground: palette.neutral.tone(10),
      surface: palette.neutral.tone(99),
      onSurface: palette.neutral.tone(10),
      surfaceVariant: palette.neutralVariant.tone(90),
      onSurfaceVariant: palette.neutralVariant.tone(30),
      outline: palette.neutralVariant.tone(50),
      outlineVariant: palette.neutralVariant.tone(80),
      shadow: palette.neutral.tone(0),
      scrim: palette.neutral.tone(0),
      inverseSurface: palette.neutral.tone(20),
      inverseOnSurface: palette.neutral.tone(95),
      inversePrimary: palette.primary.tone(80)
    )
  }

  static func darkFromCorePalette(_ palette: CorePalette) -> Scheme {
    return Scheme(
      primary: palette.primary.tone(80),
      onPrimary: palette.primary.tone(20),
      primaryContainer: palette.primary.tone(30),
      onPrimaryContainer: palette.primary.tone(90),
      secondary: palette.secondary.tone(80),
      onSecondary: palette.secondary.tone(20),
      secondaryContainer: palette.secondary.tone(30),
      onSecondaryContainer: palette.secondary.tone(90),
      tertiary: palette.tertiary.tone(80),
      onTertiary: palette.tertiary.tone(20),
      tertiaryContainer: palette.tertiary.tone(30),
      onTertiaryContainer: palette.tertiary.tone(90),
      error: palette.error.tone(80),
      onError: palette.error.tone(20),
      errorContainer: palette.error.tone(30),
      onErrorContainer: palette.error.tone(80),
      background: palette.neutral.tone(10),
      onBackground: palette.neutral.tone(90),
      surface: palette.neutral.tone(10),
      onSurface: palette.neutral.tone(90),
      surfaceVariant: palette.neutralVariant.tone(30),
      onSurfaceVariant: palette.neutralVariant.tone(80),
      outline: palette.neutralVariant.tone(60),
      outlineVariant: palette.neutralVariant.tone(30),
      shadow: palette.neutral.tone(0),
      scrim: palette.neutral.tone(0),
      inverseSurface: palette.neutral.tone(90),
      inverseOnSurface: palette.neutral.tone(20),
      inversePrimary: palette.primary.tone(40)
    )
  }
}

File: go/scheme/scheme.go
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package scheme provides Material Design color scheme generation.
package scheme

import (
	"github.com/material-foundation/material-color-utilities/go/palettes"
)

// Scheme represents a Material color scheme, a mapping of color roles to colors.
type Scheme struct {
	Primary, OnPrimary                               uint32
	PrimaryContainer, OnPrimaryContainer             uint32
	Secondary, OnSecondary                           uint32
	SecondaryContainer, OnSecondaryContainer         uint32
	Tertiary, OnTertiary                             uint32
	TertiaryContainer, OnTertiaryContainer           uint32
	Error, OnError                                   uint32
	ErrorContainer, OnErrorContainer                 uint32
	Background, OnBackground                         uint32
	Surface, OnSurface                               uint32
	SurfaceVariant, OnSurfaceVariant                 uint32
	Outline, OutlineVariant                          uint32
	Shadow, Scrim                                    uint32
	InverseSurface, InverseOnSurface, InversePrimary uint32
}

// Light creates a light theme Scheme from a source color in ARGB.
func Light(argb uint32) *Scheme {
	return lightFromCorePalette(palettes.Of(argb))
}

// Dark creates a dark theme Scheme from a source color in ARGB.
func Dark(argb uint32) *Scheme {
	return darkFromCorePalette(palettes.Of(argb))
}

// LightContent creates a light theme content-based Scheme from a source color in ARGB.
func LightContent(argb uint32) *Scheme {
	return lightFromCorePalette(palettes.ContentOf(argb))
}

// DarkContent creates a dark theme content-based Scheme from a source color in ARGB.
func DarkContent(argb uint32) *Scheme {
	return darkFromCorePalette(palettes.ContentOf(argb))
}

// lightFromCorePalette creates a light scheme from a core palette.
func lightFromCorePalette(core *palettes.Core) *Scheme {
	return &Scheme{
		Primary:              core.A1.Tone(40),
		OnPrimary:            core.A1.Tone(100),
		PrimaryContainer:     core.A1.Tone(90),
		OnPrimaryContainer:   core.A1.Tone(10),
		Secondary:            core.A2.Tone(40),
		OnSecondary:          core.A2.Tone(100),
		SecondaryContainer:   core.A2.Tone(90),
		OnSecondaryContainer: core.A2.Tone(10),
		Tertiary:             core.A3.Tone(40),
		OnTertiary:           core.A3.Tone(100),
		TertiaryContainer:    core.A3.Tone(90),
		OnTertiaryContainer:  core.A3.Tone(10),
		Error:                core.Error.Tone(40),
		OnError:              core.Error.Tone(100),
		ErrorContainer:       core.Error.Tone(90),
		OnErrorContainer:     core.Error.Tone(10),
		Background:           core.N1.Tone(99),
		OnBackground:         core.N1.Tone(10),
		Surface:              core.N1.Tone(99),
		OnSurface:            core.N1.Tone(10),
		SurfaceVariant:       core.N2.Tone(90),
		OnSurfaceVariant:     core.N2.Tone(30),
		Outline:              core.N2.Tone(50),
		OutlineVariant:       core.N2.Tone(80),
		Shadow:               core.N1.Tone(0),
		Scrim:                core.N1.Tone(0),
		InverseSurface:       core.N1.Tone(20),
		InverseOnSurface:     core.N1.Tone(95),
		InversePrimary:       core.A1.Tone(80),
	}
}

// darkFromCorePalette creates a dark scheme from a core palette.
func darkFromCorePalette(core *palettes.Core) *Scheme {
	return &Scheme{
		Primary:              core.A1.Tone(80),
		OnPrimary:            core.A1.Tone(20),
		PrimaryContainer:     core.A1.Tone(30),
		OnPrimaryContainer:   core.A1.Tone(90),
		Secondary:            core.A2.Tone(80),
		OnSecondary:          core.A2.Tone(20),
		SecondaryContainer:   core.A2.Tone(30),
		OnSecondaryContainer: core.A2.Tone(90),
		Tertiary:             core.A3.Tone(80),
		OnTertiary:           core.A3.Tone(20),
		TertiaryContainer:    core.A3.Tone(30),
		OnTertiaryContainer:  core.A3.Tone(90),
		Error:                core.Error.Tone(80),
		OnError:              core.Error.Tone(20),
		ErrorContainer:       core.Error.Tone(30),
		OnErrorContainer:     core.Error.Tone(80),
		Background:           core.N1.Tone(10),
		OnBackground:         core.N1.Tone(90),
		Surface:              core.N1.Tone(10),
		OnSurface:            core.N1.Tone(90),
		SurfaceVariant:       core.N2.Tone(30),
		OnSurfaceVariant:     core.N2.Tone(80),
		Outline:              core.N2.Tone(60),
		OutlineVariant:       core.N2.Tone(30),
		Shadow:               core.N1.Tone(0),
		Scrim:                core.N1.Tone(0),
		InverseSurface:       core.N1.Tone(90),
		InverseOnSurface:     core.N1.Tone(20),
		InversePrimary:       core.A1.Tone(40),
	}
}
